### PR TITLE
Partially implement `URLLoader` and related classes

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR implements the `URLLoader` class, allowing AVM2 scripts
to load data from a URL. This requires several other related
classes (`URLLoaderDataFormat`, `URLRequest`, `IOError`) to be
implemented as well.

Currently implemented:
* Fetching from URLs using the 'navigator' backend
* The `text` and `binary` data formats (which store data
in a `String` or `ByteArray` respectively)
* The `open`, `complete`, and `ioError` events
* The `bytesLoaded`, `bytesTotal`, and `data` properties

Not yet implemented:
* The HTTP and security events
* All of the properties of `IOError`
* The properties on `URLRequest` (besides `url`)
* The "variables" data format

This should be enough to get some basic uses of `URLLoader` working
(e.g. simple GET requests to a particular website).

Note that in Flash's `playerglobal`, the `URLLoader` class is just
a think wrapper around the more general `URLStream`. However,
implementing `URLStream` will require changes to `Navigator``
to support notifications when data arrives in the stream. When
that happens, we should be able to re-use a large amount of the
code in this PR.

### CODE REVIEW

**Comment by Aaron1011** [comment]
I'm including the system error message in the `IOError`, which means that the output will be different on Windows. The options are:

* Add platform-specific output files to our test framework
* Remove the actual system error message from the `IOError.text`
* Don't print out `IOError.text`

File: core/src/avm2/globals/flash/net/url_loader.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+pub fn class_init<'gc>(
+    _activation: &mut Activation<'_, 'gc, '_>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    Ok(Value::Undefined)
+}
+
+/// Implements `flash.net.URLLoader`'s instance constructor.
+pub fn instance_init<'gc>(
```
Comment (adrian17) [inline core/src/avm2/globals/flash/net/url_loader.rs:25]: Method bindings don't need to be `pub` (some existing code has it unnecessarily)

File: core/src/avm2/globals/flash/net/url_loader.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut this) = this {
+        activation.super_init(this, &[])?;
+        this.set_property(
+            &QName::new(Namespace::public(), "dataFormat").into(),
+            "text".into(),
+            activation,
+        )?;
+        this.set_property(
+            &QName::new(Namespace::public(), "data").into(),
```
Comment (adrian17) [inline core/src/avm2/globals/flash/net/url_loader.rs:38]: uninitialized slot should be `undefined` by default, right? No need to set it manually, I think?

File: core/src/avm2/globals/flash/net/url_loader.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+pub fn bytes_total<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(this) = this {
+        let data =
+            this.get_property(&QName::new(Namespace::public(), "data").into(), activation)?;
```
Comment (adrian17) [inline core/src/avm2/globals/flash/net/url_loader.rs:71]: FYI we have `QName::dynamic_name` to skip the `Namespace::public()` part (...we probably should rename it to `QName::public`)

File: core/src/avm2/globals/flash/net/url_loader.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        };
+
+        let data_format = this
+            .get_property(
+                &QName::new(Namespace::public(), "dataFormat").into(),
+                activation,
+            )?
+            .coerce_to_string(activation)?;
+
+        let data_format = if data_format == AvmString::from("binary") {
```
Comment (adrian17) [inline core/src/avm2/globals/flash/net/url_loader.rs:108]: would `== b"binary"` work here?

File: core/src/avm2/globals/flash/net/url_loader.rs (hunk 2 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/net/url_loader.rs:38]: It appeared to be `null` when I tested it

Comment (adrian17) [reply core/src/avm2/globals/flash/net/url_loader.rs:38]: that's weird, looking at the code we should be default-initializing these to Undefined... I'll check later too

Comment (Aaron1011) [reply core/src/avm2/globals/flash/net/url_loader.rs]: This happens here, where we create a slot trait with no default value: https://github.com/ruffle-rs/ruffle/blob/de99dcd37b5442a87c011d603858811a1947aecd/core/src/avm2/class.rs#L755-L759

This results in us picking `null` as the default slot value: https://github.com/ruffle-rs/ruffle/blob/de99dcd37b5442a87c011d603858811a1947aecd/core/src/avm2/traits.rs#L405

Side node - is the "slot type" actually used anywhere besides picking the default value?

**Comment by danielhjacobs** [comment]
I opened #7025 which clashes with this as I started work on it before I noticed that implemented "text" on IOErrorEvent.

File: core/src/avm2/globals/flash/net/url_loader.rs (hunk 2 above)
Comment (adrian17) [reply core/src/avm2/globals/flash/net/url_loader.rs]: > Side node - is the "slot type" actually used anywhere besides picking the default value?

Pretty sure it should be used for type checks, implicit conversions? And JIT ofc

> This results in us picking null as the default slot value:

Oh, I see now, it's here:

https://github.com/ruffle-rs/ruffle/blob/de99dcd37b5442a87c011d603858811a1947aecd/core/src/avm2/traits.rs#L159

For some reason I thought the `None` gets propagated all the way to actual slot value iniitialization, up to here:
https://github.com/ruffle-rs/ruffle/blob/de99dcd37b5442a87c011d603858811a1947aecd/core/src/avm2/object/script_object.rs#L294-L298

But you're right, it's indeed `null` by default in FP.

**Review by Herschel — APPROVED** [review]
Thank you!

**Comment by Herschel** [comment]
We want to mimic Flash's error messages as close as possible, so we shouldn't include the OS error message in `IOErrorEvent.text`; for now it should say something like "Error #2032: Stream Error" so it's easy on our tests.

Eventually there is a fairly big TODO to properly implement the Error types (which use methods like `Error.formatToString` to build the error strings), create a large table of specific of error #-to-messages somewhere in our codebases, and make it easy for us to throw specific errors. Somewhat connected with #5916.

**Comment by Aaron1011** [comment]
@Herschel I've adjusted the error message

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**